### PR TITLE
[Serializer] DateTimeNormalizer: allow to provide timezone

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

My own use-case was for denormalization of a csv file provided by a third-party. The datetime format inside does not contain any timezone information, and won't change, but it's established to be UTC (or at least consistent).

So by providing the new `datetime_timezone` option, the returned instance of `\DateTime(Interface)` will properly be set with the expected timezone. (In case the format already supports the time offset, the provided timezone is ignored in favor of the one parsed by the `\DateTime` object)

Regarding normalization, the expected behavior of this feature is to consistently return the same time offset.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 1]
```diff
@@ -49,6 +57,11 @@ public function normalize($object, $format = null, array $context = array())
         }
 
         $format = isset($context[self::FORMAT_KEY]) ? $context[self::FORMAT_KEY] : $this->format;
+        $timezone = $this->getTimezone($context);
+
+        if (null !== $timezone) {
+            $object = (new \DateTime('@'.$object->getTimestamp()))->setTimezone($timezone);
```
Comment (maidmaid) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: ``new \DateTime('@'.$object->getTimestamp(), $timezone)`` instead?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 2]
```diff
@@ -69,9 +82,15 @@ public function supportsNormalization($data, $format = null)
     public function denormalize($data, $class, $format = null, array $context = array())
     {
         $dateTimeFormat = isset($context[self::FORMAT_KEY]) ? $context[self::FORMAT_KEY] : null;
+        $timezone = $this->getTimezone($context);
 
         if (null !== $dateTimeFormat) {
-            $object = \DateTime::class === $class ? \DateTime::createFromFormat($dateTimeFormat, $data) : \DateTimeImmutable::createFromFormat($dateTimeFormat, $data);
+            if (null === $timezone) {
+                // https://bugs.php.net/bug.php?id=68669
```
Comment (maidmaid) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: As [this issue](https://github.com/php/php-src/pull/1167) is related to an unsupported version PHP, I think it's not usefull to handle this problem here.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 1 above)
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: The thing is to not use the second argument here actually, because as specified on the manual entry:

> The $timezone parameter and the current timezone are ignored when the $time parameter either is a UNIX timestamp (e.g. @946684800) or specifies a timezone (e.g. 2010-01-28T15:00:00+02:00).

— http://php.net/manual/en/datetime.construct.php#refsect1-datetime.construct-parameters

 on the contrary of `\DateTime::setTimeZone()` which will properly set the timezone and update the date accordingly.

See the following:

```php
>>> new \DateTime('@1480546800')
=> DateTime {#166
     +"date": "2016-11-30 23:00:00.000000",
     +"timezone_type": 1,
     +"timezone": "+00:00",
   }
>>> new \DateTime('@1480546800', new \DateTimeZone('Japan'))
=> DateTime {#173
     +"date": "2016-11-30 23:00:00.000000",
     +"timezone_type": 1,
     +"timezone": "+00:00",
   }
>>> (new \DateTime('@1480546800'))->setTimeZone(new \DateTimeZone('Japan'))
=> DateTime {#174
     +"date": "2016-12-01 08:00:00.000000",
     +"timezone_type": 3,
     +"timezone": "Japan",
   }
>>>
```

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 2 above)
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: We can remove this workaround in Symfony 4.0, but for now the Symfony 3 [still supports PHP 5.5](https://github.com/symfony/symfony/blob/master/composer.json#L19).
See failing tests on Travis: https://travis-ci.org/symfony/symfony/jobs/222333390#L2437-L2471

Comment (maidmaid) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: In this case, maybe is it better to check PHP version like in [this example](https://github.com/symfony/symfony/blob/ac5cfee6f12f5c93fe8d8b67941fd478dfaa397a/src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php#L121-L122)?

Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Updated. Thank you.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 3]
```diff
@@ -23,18 +23,26 @@
...
+    const TIMEZONE_KEY = 'datetime_timezone';
 
     /**
      * @var string
      */
     private $format;
 
     /**
-     * @param string $format
+     * @var \DateTimeZone
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: This docblock is useless (same for the existing docblock for `$format` btw).

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: I suggest to use `DateTimeImmutable` instead of `DateTime`.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 4]
```diff
@@ -126,4 +145,20 @@ private function formatDateTimeErrors(array $errors)
...
     }
+
+    /**
+     * @param array $context
+     *
+     * @return \DateTimeZone|null
+     */
+    private function getTimezone(array $context)
+    {
+        $dateTimeZone = array_key_exists(self::TIMEZONE_KEY, $context) ? $context[self::TIMEZONE_KEY] : $this->timezone;
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: You can use `isset` instead of `array_key_exists` here (faster as it's not a function call).

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 4 above)
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: The idea behind this was to allow setting `null` as timezone, in case one was provided in the constructor.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 3 above)
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Oh. I misread. I though you were talking about the private `DateTimeNormalizer::getTimezone()`. 
While I agree, I added it to be consistent with the existing docblock. If we want to remove it, shouldn't we make a dedicated PR on the lowest branch rather than here?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 1 above)
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Done

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 3 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Indeed.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Ok got it.

**Review by dunglas — APPROVED** [review]

**Review by dunglas — APPROVED** [review]
:+1:

**Review by xabbuh — APPROVED** [review]

**Comment by ogizanagi** [comment]
Ready to go? :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ogizanagi.

**Comment by GuilhemN** [comment]
It seems this breaks sf 3.4 tests (see https://travis-ci.org/symfony/symfony/jobs/243836779). Are you sure the issue you talked about is fixed in php 5.6?

**Comment by ogizanagi** [comment]
Hmm, indeed, it seems I misinterpreted https://github.com/php/php-src/pull/1167#issuecomment-270340340. Considering https://3v4l.org/NmFK5 and the list of tags containing https://github.com/php/php-src/commit/365f428d463502fd8a2c63640b280753df4bf71f, it seems it's only available since PHP 7...

Thanks for the report. See https://github.com/symfony/symfony/pull/23217.
